cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java

```diff
@@ -428,8 +428,20 @@ private boolean canAppearNext(Property prop) {
                 final List<Property> propList = reqProps.asPropertyList();
                 result = new ArrayList<CommandArgument>(propList.size());
 
-                String radical = ctx.getParsedCommandLine().getLastParsedPropertyValue() == null
-                        ? ctx.getParsedCommandLine().getLastParsedPropertyName() : null;
+                String radical = null;
+                if(ctx.getParsedCommandLine().getLastParsedPropertyValue() == null) {
+                    radical = ctx.getParsedCommandLine().getLastParsedPropertyName();
+                    //Check if the property is completely specified and is negated
```

Comment:
Why do you need to set the radical to null in this case? Would you have an example?

Reply:
The methond canAppearNext() [1] returns false when the property is already in the input line or when the name does not start with radical. That means if the operation e.g. :reload-servers has 2 possible properties "blocking" and "start-mode" then in case the "!blocking" is specified the radical would be set to "blocking" and the  canAppearNext() would return false for every one command, since the "blocking" is already  in the input and the other properties don't start with "blocking" string.

In the old code the radical was set to null for example in this case ":reload-servers(blocking=false<TAB>" because the last property blocking has the value false and getLastParsedPropertyValue() returned "false". 
However, In case  ":reload-servers(!blocking<TAB>" which has similar meaning as the first one, the radical was set to "blocking". This was caused because the getLastParsedPropertyValue() returned null which I think is right, since there is no explicit specified value like in the first case the "=false". Therefore when the radical was set to null in the first case I think we could do the same for the case when the property is negated and its name is completely specified e.g. "!blocking".

[1] https://github.com/MMarus/wildfly-core/blob/11911af7eea6dfd89c8cb2f843524a8c75230282/cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java#L378

Reply:
Thank-you for taking the time to detail the case. You are right, you perhaps have to take into account a corner case, let's imagine we have blocking and blocking-param options.
reload:(!blocking<TAB> ==> should propose "," and blocking-param. This is something that you should be able to reproduce in your new cli test.
BTW, when making changes to completion it is good to extend the test https://github.com/wildfly/wildfly-core/blob/master/testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/management/cli/CliCompletionTestCase.java

Reply:
I just looked into the test *PropertiesCompletionTestCase* again and the *DefaultOperationCandidatesProvider* is not tested there, since there is a *MockOperationCandidatesProvider* used instead. I think it's because the method *getProperties()* first creates and executes request in order to get list of properties which wouldn't be easy to mock. However, in order to create the resulting list of  *CommandArgument* only the *Property* list is needed (line 429-524 [1]) which could be very easily moved to separate method and it would allow to test this functionality.

Do You think I can move that part into separate method or is it too big change ?
Thank You for answer.

[1] https://github.com/MMarus/wildfly-core/blob/80f7c778602d3be0b05d0cc89d20a7efd4c89c0b/cli/src/main/java/org/jboss/as/cli/operation/impl/DefaultOperationCandidatesProvider.java#L429

Reply:
@MMarus , +1 to increase unit-testability of completion!